File: 03/gpt-5.1/typescript/solution.ts
import * as fs from 'fs';

const lines = fs.readFileSync('input.txt', 'utf-8').split(/\r?\n/);

let totalJoltagePart1 = 0;
let totalJoltagePart2 = 0;

for (const line of lines) {
	if (!line.trim()) continue;

	let maxJoltagePart1 = 0;
	let found = false;

	for (let d1 = 9; d1 >= 1 && !found; d1--) {
		const c1 = String(d1);
		const idx1 = line.indexOf(c1);
		if (idx1 === -1) continue;

		for (let d2 = 9; d2 >= 1; d2--) {
			const c2 = String(d2);
			const idx2 = line.lastIndexOf(c2);
			if (idx2 === -1) continue;

			if (idx1 < idx2) {
				maxJoltagePart1 = d1 * 10 + d2;
				found = true;
				break;
			}
		}
	}

	totalJoltagePart1 += maxJoltagePart1;

	// Part 2: choose exactly 12 digits to form the largest possible number
	const digits = line;
	const targetLength = 12;
	const chosen: string[] = [];
	let start = 0;

	for (let pos = 0; pos < targetLength; pos++) {
		const remainingNeeded = targetLength - pos - 1;
		let bestDigit: string | null = null;
		let bestIndex = -1;

		for (let d = 9; d >= 0; d--) {
			const ch = String(d);
			let lastIndex = -1;
			for (let i = start; i < digits.length; i++) {
				if (digits[i] === ch) {
					lastIndex = i;
				}
			}
			if (lastIndex < start) continue;

			if (lastIndex <= digits.length - 1 - remainingNeeded) {
				bestDigit = ch;
				bestIndex = lastIndex;
				break;
			}
		}

		if (bestIndex === -1) {
			bestDigit = digits[start];
			bestIndex = start;
		}

		chosen.push(bestDigit!);
		start = bestIndex + 1;
	}

	let lineJoltagePart2 = 0;
	for (const ch of chosen) {
		lineJoltagePart2 = lineJoltagePart2 * 10 + Number(ch);
	}

	totalJoltagePart2 += lineJoltagePart2;
}

console.log(`Day 03 Part 1: ${totalJoltagePart1}`);
console.log(`Day 03 Part 2: ${totalJoltagePart2}`);

File: 03/gpt-5.1/python/solution.py
total_joltage_part1 = 0
total_joltage_part2 = 0

with open("input.txt") as f:
    for line in f:
        line = line.strip()
        if not line:
            continue

        max_joltage_part1 = 0
        found = False

        for d1 in range(9, 0, -1):
            c1 = str(d1)
            idx1 = line.find(c1)
            if idx1 == -1:
                continue

            for d2 in range(9, 0, -1):
                c2 = str(d2)
                idx2 = line.rfind(c2)
                if idx2 == -1:
                    continue

                if idx1 < idx2:
                    max_joltage_part1 = d1 * 10 + d2
                    found = True
                    break

            if found:
                break

        total_joltage_part1 += max_joltage_part1

        # Part 2: choose exactly 12 digits to form the largest possible number
        digits = line
        target_length = 12
        chosen = []
        start = 0

        for pos in range(target_length):
            remaining_needed = target_length - pos - 1
            best_digit = None
            best_index = -1

            for d in range(9, -1, -1):
                ch = str(d)
                last_index = digits.rfind(ch, start)
                if last_index == -1:
                    continue

                if last_index <= len(digits) - 1 - remaining_needed:
                    best_digit = ch
                    best_index = last_index
                    break

            if best_index == -1:
                best_digit = digits[start]
                best_index = start

            chosen.append(best_digit)
            start = best_index + 1

        line_joltage_part2 = int("".join(chosen))
        total_joltage_part2 += line_joltage_part2

print(f"Day 03 Part 1: {total_joltage_part1}")
print(f"Day 03 Part 2: {total_joltage_part2}")
